core/src/main/java/org/htsjdk/core/utils/PathSpecifier.java

```diff
@@ -0,0 +1,238 @@
+package org.htsjdk.core.utils;
+
+import org.htsjdk.core.api.PathURI;
+import org.htsjdk.core.exception.HtsjdkException;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.io.Serializable;
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.nio.file.*;
+import java.nio.file.spi.FileSystemProvider;
+
+/**
+ * Default implementation for PathURI.
+ *
+ * This class takes a raw string that is to be interpreted as a path specifier, and converts it internally to a URI or
+ * Path object. If no scheme is provided as part of the raw string used in the constructor(s), the URI is assumed
+ * to represent a file on the local file system, and will be backed by a URI with a "file:/" scheme and a path
+ * part that is automatically encoded/escaped to ensure it is a valid URI. If the raw string contains a scheme,
+ * it will be backed by a URI formed from the raw string as presented, with no further encoding/escaping.
+ *
+ * For example, a URI that contains a scheme, and has an embedded "#" in the path will be treated as having a
+ * fragment delimiter. If the URI contains an scheme, the "#" will be escaped.
+ *
+ * There are 3 succeeding levels of validation:
+ *
+ * 1) PathSpecifier constructor - requires a syntactically valid URI, possibly containing a scheme (if no scheme
+ * is present the path part will be escaped/encoded)
+ * 2) isNio - returns true if the URI is syntactically valid, and there is an installed NIO provider that matches
+ * the URI scheme
+ * 3) isPath - syntactically valid URI that can be resolved to a java.io.Path by the provider
+ *
+ * <scheme>:<scheme-specific-part>
+ * <scheme>://<authority><path>?<query>
+ *  absoluteURI   = scheme ":" ( hier_part | opaque_part )
+ *       hier_part     = ( net_path | abs_path ) [ "?" query ]
+ *       net_path      = "//" authority [ abs_path ]
+ *       abs_path      = "/"  path_segments
+ *
+ * A URI is absolute if, and only if, it has a scheme component.
+ * A URI is opaque if, and only if, it is absolute (has a scheme) and its
+ * scheme-specific part does not begin with a slash character ('/')
+ *
+ * A relative reference that does not begin with a scheme name or a
+ * slash character is termed a relative-path reference.
+ *
+ * A relative reference beginning with a single slash character is
+ * termed an absolute-path reference, as defined by <abs_path> in
+ * Section 3.
+ *
+ * URI that do not make use of the slash "/" character for separating
+ * hierarchical components are considered opaque by the generic URI
+ * parser.
+ */
+public class PathSpecifier implements PathURI, Serializable {
```

Comment:
Probably we should make our defaults implementations final, unless required. The same as before, we should not allow to extend classes unless we explicitly see a widely spread use-case.

Reply:
Also, should `Serializable` be part of the `PathURI` API?

Reply:
Good question on `Serializable`. As for `final`, the intent is to provide a default implementation that allows overrides with custom resolution. GATK will certainly need to do this for gcs, etc.